sdk/core/azure-core/test/ut/websocket_test.cpp

```diff
@@ -0,0 +1,875 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// SPDX-License-Identifier: MIT
+
+#include "../../src/http/websockets/websockets_impl.hpp"
+#include "azure/core/http/websockets/websockets.hpp"
+#include "azure/core/internal/json/json.hpp"
+#include <chrono>
+#include <gtest/gtest.h>
+#include <list>
+#include <set>
+#include <thread>
+#if defined(BUILD_CURL_HTTP_TRANSPORT_ADAPTER)
+#include "azure/core/http/websockets/curl_websockets_transport.hpp"
+#endif
+// cspell::words closeme flibbityflobbidy
+
+using namespace Azure::Core;
+using namespace Azure::Core::Http::WebSockets;
+using namespace Azure::Core::Http::WebSockets::_internal;
+using namespace std::chrono_literals;
+
+constexpr uint16_t UndefinedButLegalCloseReason = 4500;
+
+class WebSocketTests : public testing::Test {
+private:
+protected:
+  // Create
+  static void SetUpTestSuite() {}
+  static void TearDownTestSuite() {}
+};
+
+TEST_F(WebSocketTests, CreateSimpleSocket)
+{
+  {
+    WebSocket defaultSocket(Azure::Core::Url("http://localhost:8000"));
+    defaultSocket.AddHeader("newHeader", "headerValue");
+    EXPECT_THROW(defaultSocket.GetNegotiatedProtocol(), std::runtime_error);
+  }
+}
+
+TEST_F(WebSocketTests, OpenSimpleSocket)
+{
+  {
+    WebSocketOptions options;
+    WebSocket defaultSocket(Azure::Core::Url("http://localhost:8000/openclosetest"), options);
+    defaultSocket.AddHeader("newHeader", "headerValue");
+
+    defaultSocket.Open();
+
+    EXPECT_THROW(defaultSocket.AddHeader("newHeader", "headerValue"), std::runtime_error);
+
+    // Close the socket without notifying the peer.
+    defaultSocket.Close();
+  }
+
```

Comment:
Would be nice to add a general purpose mechanism into tests to validate we don't leak handles.

Reply:
Fair. In the case of the WebSocket tests, the `Close` method actually engages with the server.

But handle leak tests (and memory leak tests) are a good idea. Possibly similar to the stress tests?